File: 2807/Solution.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var insertGreatestCommonDivisors = function(head) {
    let prev = head;
    let cur = head.next;
    const findGcd = (a, b) => {
        while(b != 0){
            let temp = a % b;
            a = b;
            b = temp;
        }
        return a;
    }
    while(cur != null){
        let gcd = findGcd(prev.val, cur.val);
        prev.next = new ListNode(gcd, cur);
        prev = cur;
        cur = cur.next;
    } 
    return head;   

};

File: 2807/Solution.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func insertGreatestCommonDivisors(head *ListNode) *ListNode {
    prev, cur := head, head.Next

    for cur != nil {
        gcd := findGcd(prev.Val, cur.Val)
        prev.Next = &ListNode{Val: gcd, Next: cur}
        prev = cur
        cur = cur.Next
    }
    return head
}
func findGcd(a, b int) int {
    for b != 0 {
        temp := a % b
        a = b
        b = temp
    }
    return a
}
